sytest-whitelist

```diff
@@ -123,6 +123,7 @@ We should see our own leave event when rejecting an invite, even if history_visi
 Newly left rooms appear in the leave section of gapped sync
 Previously left rooms don't appear in the leave section of sync
 Left rooms appear in the leave section of full state sync
+Archived rooms only contain history from before the user left
```

Comment:
For the test with Dendrite, I am seeing this behavior for [`tests/31sync/09archived.pl#L371-L372`](https://github.com/matrix-org/sytest/blob/11240a80c42994a3de87f6e16b9452dacff2f2c4/tests/31sync/09archived.pl#L371-L372). Only the proper 2 events are shown but the `before` timeline event is second instead of first which Sytest expects. At this point, I am trying to run Sytest against Synapse to see how that behaves. Maybe just an ordering thing.

```sh
$ POSTGRES=1 ./run-tests.pl tests/31sync/09archived.pl -I Dendrite::Monolith -d ../dendrite/bin -W ../dendrite/sytest-whitelist  --stop-on-fail
 Testing if: Archived rooms only contain history from before the user left... FAIL:
 | Expected only events from before leaving at tests/31sync/09archived.pl line 373.
...
# 0.430608: Returned sync body
 # {
 #   account_data => {
 #     events => [
 #       {
 #         content => {
 #           global => { content => [], override => [], room => [], sender => [], underride => [] },
 #         },
 #         type => "m.push_rules",
 #       },
 #     ],
 #   },
 #   device_lists => {
 #     changed => ["\@anon-20201230_075726-11:localhost:8800"],
 #     left => ["\@anon-20201230_075726-10:localhost:8800"],
 #   },
 #   device_one_time_keys_count => {},
 #   next_batch => "s101_0_0_0.dl-0-12",
 #   presence => { events => [] },
 #   rooms => {
 #     invite => {},
 #     join   => {},
 #     leave  => {
 #                 "!FCop8FToKXYtNSk8:localhost:8800" => {
 #                   state => { events => [] },
 #                   timeline => {
 #                     events     => [
 #                                     {
 #                                       content          => {
 #                                                             avatar_url  => "",
 #                                                             displayname => "anon-20201230_075726-11",
 #                                                             membership  => "join",
 #                                                           },
 #                                       event_id         => "\$nwrV5xShxLY8wEDP7ptEQITa0QE-LWgMM7hrgcpa44k",
 #                                       origin_server_ts => JSON::number(1609315050339),
 #                                       sender           => "\@anon-20201230_075726-11:localhost:8800",
 #                                       state_key        => "\@anon-20201230_075726-11:localhost:8800",
 #                                       type             => "m.room.member",
 #                                       unsigned         => {},
 #                                     },
 #                                     {
 #                                       content => { body => "before", msgtype => "m.text" },
 #                                       event_id => "\$bNSkFZt8OfEcnQv5No2QOMBbSE0SC0Vjs01Zs3EPGZM",
 #                                       origin_server_ts => JSON::number(1609315050381),
 #                                       sender => "\@anon-20201230_075726-10:localhost:8800",
 #                                       type => "m.room.message",
 #                                     },
 #                                   ],
 #                     limited    => JSON::false,
 #                     prev_batch => "t5_1096",
 #                   },
 #                 },
 #               },
 #     peek   => {},
 #   },
 #   to_device => { events => [] },
 # }
 ```

Reply:
Running into a lot of issues trying to run Sytest against Synapse as well. Added some PR's for some basic things

 - https://github.com/matrix-org/synapse/pull/8997
 - https://github.com/matrix-org/sytest/pull/994
 - https://github.com/matrix-org/sytest/pull/995

---

Next blocker is that I am seeing `Cannot bind - No such file or directory` (`@details=bind IO::Socket::UNIX=GLOB(0x7fd150810318) #sytest-LGNDECQIXTSOSTLHPTKI.sock No such file or directory at tests/05homeserver.pl line 131.`). The perl isn't making this easy to debug and the stack trace isn't very helpful when it's pointing to where the error is caught only.

The socket name `#sytest-LGNDECQIXTSOSTLHPTKI.sock` is from [`lib/SyTest/Homeserver/ProcessManager.pm#L323`](https://github.com/matrix-org/sytest/blob/11240a80c42994a3de87f6e16b9452dacff2f2c4/lib/SyTest/Homeserver/ProcessManager.pm#L323) and looks like the binding is done there also. Need to look at that more.


```sh
$ POSTGRES=1 ./run-tests.pl tests/31sync/09archived.pl -C -S
Running tests/31sync/09archived.pl...
 # Presuming ability 'can_sync'
  Testing if: Left rooms appear in the leave section of sync...
Tests: 0 / 8 | OKSignature ok
subject=/CN=localhost
 # Started test HTTPS Server at localhost:56641
  Testing if: Left rooms appear in the leave section of sync...
Tests: 0 / 8 | OKRequesting:
  POST https://localhost:56641/http_server_self_test HTTP/1.1
  Host: localhost
  Content-Length: 25
  Content-Type: application/json

  {"some_key":"some_value"}
 # Started test SMTP Server at localhost:56642
  Testing if: Left rooms appear in the leave section of sync...
Tests: 0 / 8 | OKReceived Request for POST /http_server_self_test:
  {"some_key":"some_value"}
--
Sending response to /http_server_self_test:
  200 OK
  Content-Length: 2
  Content-Type: application/json

  {}
--
Response from POST /http_server_self_test:
  HTTP/1.1 200 OK
  Content-Length: 2
  Content-Type: application/json

  {}
--
Requesting:
  POST https://localhost:56641/http_server_self_test HTTP/1.1
  Host: localhost
  Content-Length: 2
  Content-Type: application/json

  {}
--
Received Request for POST /http_server_self_test:
  {}
--
Sending response to /http_server_self_test:
  200 OK
  Content-Length: 33
  Content-Type: application/json

  {"response_key":"response_value"}
--
Response from POST /http_server_self_test:
  HTTP/1.1 200 OK
  Content-Length: 33
  Content-Type: application/json

  {"response_key":"response_value"}
 # Starting Homeserver using SyTest::HomeserverFactory::Synapse=HASH(0x7fd14b00af90)
 # Starting server-0
 # Using DB config from /Users/eric/Documents/github/sytest/server-0/database.yaml
 # Clearing SQLite database at :memory:
  Testing if: Left rooms appear in the leave section of sync...
Tests: 0 / 8 | OKSignature ok
subject=/CN=localhost
 # Generating config for port 8800
 # Creating config for server 0 with command python -m synapse.app.homeserver --config-path /Users/eric/Documents/github/sytest/server-0/config.yaml --generate-config --report-stats=no --server-name localhost:8800
 # Starting server 0 for port 8800
  Testing if: Left rooms appear in the leave section of sync...
Tests: 0 / 8 | OKStarting synapse with command: python -m synapse.app.homeserver --config-path /Users/eric/Documents/github/sytest/server
 ** Error starting server-0 (on port 8800): Cannot bind - No such file or directory @details=bind IO::Socket::UNIX=GLOB(0x7fd150810318) #sytest-LGNDECQIXTSOSTLHPTKI.sock No such file or directory at tests/05homeserver.pl line 131.
 ** ServerInfo: https://localhost:8800, localhost, localhost:8800 8800 at tests/05homeserver.pl line 132.
  Testing if: Left rooms appear in the leave section of sync...
Tests: 0 / 8 | OK
 # Killing process server
 ** server process failed - code=2
  Testing if: Left rooms appear in the leave section of sync...
Tests: 0 / 8 | OK%
```

Reply:
It's an abstract socket starting with a `\0` null byte. More info, https://unix.stackexchange.com/questions/206386/what-does-the-symbol-denote-in-the-beginning-of-a-unix-domain-socket-path-in-l

I'm not sure why I am seeing these `Cannot bind - No such file or directory` errors for the socket which shouldn't be on the file system at all.

Reply:
I confirmed behavior on Synapse is what the test expects `before` as the first timeline event. Either the ordering on the `timeline` events should be most recent first(and I need to update my implementation) or the Sytest with Synapse just has a race condition where the membership event shows up after the other one.

The spec doesn't make it clear what the ordering on the `timeline` events are, https://matrix.org/docs/spec/client_server/r0.6.1#get-matrix-client-r0-sync

---


For reference, I used `print STDERR "\nReturned sync body $body" . Dumper($body) . "\n";` to get it to show up in the Docker Synapse results. For Dendrite, I could use `log_if_fail "Returned sync body", $body;` because the test would fail.

```
$ docker run --rm --name sytest -v "/Users/eric/Documents/github/sytest:/sytest" \
  -v "/Users/eric/Documents/github/synapse:/src" \
  -v "/Users/eric/logs:/logs" \
  -v "/Users/eric/go:/gopath" -e "POSTGRES=1" \
  matrixdotorg/sytest-synapse:py37 tests/31sync/09archived.pl

...

 Test 8 Archived rooms only contain history from before the user left...
Returned sync body HASH(0x55dae2783e98)$VAR1 = {
    'device_lists' => {
            'left' => [],
            'changed' => []
          },
    'account_data' => {
            'events' => [
              {
                'type' => 'm.push_rules',
                'content' => {
                   'global' => {
                     'underride' => [
                          {
                            'conditions' => [
                            {
                              'kind' => 'event_match',
                              'pattern' => 'm.call.invite',
                              'key' => 'type'
                            }
                                ],
                            'actions' => [
                               'notify',
                               {
                                 'set_tweak' => 'sound',
                                 'value' => 'ring'
                               },
                               {
                                 'value' => bless( do{\(my $o = 0)}, 'JSON::PP::Boolean' ),
                                 'set_tweak' => 'highlight'
                               }
                             ],
                            'enabled' => bless( do{\(my $o = 1)}, 'JSON::PP::Boolean' ),
                            'rule_id' => '.m.rule.call',
                            'default' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'}
                          },
                          {
                            'conditions' => [
                            {
                              'kind' => 'room_member_count',
                              'is' => '2'
                            },
                            {
                              'kind' => 'event_match',
                              'pattern' => 'm.room.message',
                              'key' => 'type'
                            }
                                ],
                            'actions' => [
                               'notify',
                               {
                                 'value' => 'default',
                                 'set_tweak' => 'sound'
                               },
                               {
                                 'value' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'actions'}[2]{'value'},
                                 'set_tweak' => 'highlight'
                               }
                             ],
                            'default' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'},
                            'rule_id' => '.m.rule.room_one_to_one',
                            'enabled' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'}
                          },
                          {
                            'enabled' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'},
                            'rule_id' => '.m.rule.encrypted_room_one_to_one',
                            'default' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'},
                            'conditions' => [
                            {
                              'kind' => 'room_member_count',
                              'is' => '2'
                            },
                            {
                              'key' => 'type',
                              'pattern' => 'm.room.encrypted',
                              'kind' => 'event_match'
                            }
                                ],
                            'actions' => [
                               'notify',
                               {
                                 'set_tweak' => 'sound',
                                 'value' => 'default'
                               },
                               {
                                 'set_tweak' => 'highlight',
                                 'value' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'actions'}[2]{'value'}
                               }
                             ]
                          },
                          {
                            'conditions' => [
                            {
                              'key' => 'type',
                              'kind' => 'event_match',
                              'pattern' => 'm.room.message'
                            }
                                ],
                            'actions' => [
                               'notify',
                               {
                                 'value' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'actions'}[2]{'value'},
                                 'set_tweak' => 'highlight'
                               }
                             ],
                            'default' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'},
                            'rule_id' => '.m.rule.message',
                            'enabled' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'}
                          },
                          {
                            'default' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'},
                            'rule_id' => '.m.rule.encrypted',
                            'enabled' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'},
                            'conditions' => [
                            {
                              'pattern' => 'm.room.encrypted',
                              'kind' => 'event_match',
                              'key' => 'type'
                            }
                                ],
                            'actions' => [
                               'notify',
                               {
                                 'set_tweak' => 'highlight',
                                 'value' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'actions'}[2]{'value'}
                               }
                             ]
                          },
                          {
                            'rule_id' => '.im.vector.jitsi',
                            'default' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'},
                            'enabled' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'},
                            'conditions' => [
                            {
                              'pattern' => 'im.vector.modular.widgets',
                              'kind' => 'event_match',
                              'key' => 'type'
                            },
                            {
                              'kind' => 'event_match',
                              'pattern' => 'jitsi',
                              'key' => 'content.type'
                            },
                            {
                              'key' => 'state_key',
                              'pattern' => '*',
                              'kind' => 'event_match'
                            }
                                ],
                            'actions' => [
                               'notify',
                               {
                                 'value' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'actions'}[2]{'value'},
                                 'set_tweak' => 'highlight'
                               }
                             ]
                          }
                        ],
                     'room' => [],
                     'override' => [
                         {
                           'default' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'},
                           'rule_id' => '.m.rule.master',
                           'enabled' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'actions'}[2]{'value'},
                           'conditions' => [],
                           'actions' => [
                              'dont_notify'
                            ]
                         },
                         {
                           'conditions' => [
                                 {
                             'kind' => 'event_match',
                             'pattern' => 'm.notice',
                             'key' => 'content.msgtype'
                                 }
                               ],
                           'actions' => [
                              'dont_notify'
                            ],
                           'enabled' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'},
                           'rule_id' => '.m.rule.suppress_notices',
                           'default' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'}
                         },
                         {
                           'default' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'},
                           'rule_id' => '.m.rule.invite_for_me',
                           'enabled' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'},
                           'actions' => [
                              'notify',
                              {
                                'set_tweak' => 'sound',
                                'value' => 'default'
                              },
                              {
                                'set_tweak' => 'highlight',
                                'value' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'actions'}[2]{'value'}
                              }
                            ],
                           'conditions' => [
                                 {
                             'kind' => 'event_match',
                             'pattern' => 'm.room.member',
                             'key' => 'type'
                                 },
                                 {
                             'pattern' => 'invite',
                             'kind' => 'event_match',
                             'key' => 'content.membership'
                                 },
                                 {
                             'kind' => 'event_match',
                             'pattern' => '@anon-20201231_100508-11:localhost:8800',
                             'key' => 'state_key'
                                 }
                               ]
                         },
                         {
                           'default' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'},
                           'rule_id' => '.m.rule.member_event',
                           'enabled' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'},
                           'actions' => [
                              'dont_notify'
                            ],
                           'conditions' => [
                                 {
                             'pattern' => 'm.room.member',
                             'kind' => 'event_match',
                             'key' => 'type'
                                 }
                               ]
                         },
                         {
                           'conditions' => [
                                 {
                             'kind' => 'contains_display_name'
                                 }
                               ],
                           'actions' => [
                              'notify',
                              {
                                'set_tweak' => 'sound',
                                'value' => 'default'
                              },
                              {
                                'set_tweak' => 'highlight'
                              }
                            ],
                           'enabled' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'},
                           'rule_id' => '.m.rule.contains_display_name',
                           'default' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'}
                         },
                         {
                           'actions' => [
                              'notify',
                              {
                                'value' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'},
                                'set_tweak' => 'highlight'
                              }
                            ],
                           'conditions' => [
                                 {
                             'key' => 'content.body',
                             'pattern' => '@room',
                             'kind' => 'event_match'
                                 },
                                 {
                             'key' => 'room',
                             'kind' => 'sender_notification_permission'
                                 }
                               ],
                           'default' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'},
                           'rule_id' => '.m.rule.roomnotif',
                           'enabled' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'}
                         },
                         {
                           'conditions' => [
                                 {
                             'pattern' => 'm.room.tombstone',
                             'kind' => 'event_match',
                             'key' => 'type'
                                 },
                                 {
                             'key' => 'state_key',
                             'kind' => 'event_match',
                             'pattern' => ''
                                 }
                               ],
                           'actions' => [
                              'notify',
                              {
                                'set_tweak' => 'highlight',
                                'value' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'}
                              }
                            ],
                           'rule_id' => '.m.rule.tombstone',
                           'default' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'},
                           'enabled' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'}
                         },
                         {
                           'conditions' => [
                                 {
                             'key' => 'type',
                             'pattern' => 'm.reaction',
                             'kind' => 'event_match'
                                 }
                               ],
                           'actions' => [
                              'dont_notify'
                            ],
                           'enabled' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'},
                           'rule_id' => '.m.rule.reaction',
                           'default' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'}
                         }
                       ],
                     'sender' => [],
                     'content' => [
                        {
                          'actions' => [
                             'notify',
                             {
                               'value' => 'default',
                               'set_tweak' => 'sound'
                             },
                             {
                               'set_tweak' => 'highlight'
                             }
                           ],
                          'enabled' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'},
                          'pattern' => 'anon-20201231_100508-11',
                          'rule_id' => '.m.rule.contains_user_name',
                          'default' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'}
                        }
                      ]
                   },
                   'device' => {}
                 }
              }
            ]
          },
    'to_device' => {
         'events' => []
       },
    'device_one_time_keys_count' => {},
    'org.matrix.msc2732.device_unused_fallback_key_types' => [],
    'presence' => {
        'events' => [
          {
            'sender' => '@anon-20201231_100508-11:localhost:8800',
            'type' => 'm.presence',
            'content' => {
               'presence' => 'online',
               'currently_active' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'enabled'},
               'last_active_ago' => bless( do{\(my $o = 14)}, 'JSON::number' )
                   }
          }
              ]
      },
    'groups' => {
      'leave' => {},
      'invite' => {},
      'join' => {}
          },
    'rooms' => {
           'join' => {},
           'invite' => {},
           'leave' => {
            '!hameQsXnbMgFFhAVFu:localhost:8800' => {
                        'account_data' => {
                          'events' => []
                        },
                        'state' => {
                         'events' => []
                       },
                        'timeline' => {
                            'prev_batch' => 's97_10_0_1_1_1_1_12_1',
                            'limited' => $VAR1->{'account_data'}{'events'}[0]{'content'}{'global'}{'underride'}[0]{'actions'}[2]{'value'},
                            'events' => [
                              {
                                'content' => {
                                   'msgtype' => 'm.text',
                                   'body' => 'before'
                                 },
                                'origin_server_ts' => bless( do{\(my $o = '1609409119445')}, 'JSON::number' ),
                                'type' => 'm.room.message',
                                'sender' => '@anon-20201231_100508-10:localhost:8800',
                                'event_id' => '$PLmOt32fBVWBdstlKeRSkuX9YN9uLxnaYa2y3yvPapQ',
                                'unsigned' => {
                                    'age' => bless( do{\(my $o = 287)}, 'JSON::number' )
                                  }
                              },
                              {
                                'origin_server_ts' => bless( do{\(my $o = '1609409119481')}, 'JSON::number' ),
                                'state_key' => '',
                                'event_id' => '$16fwFZwRXuSWgdRhMzhKMExm3UT2qvZfKwKoU9obpHY',
                                'sender' => '@anon-20201231_100508-10:localhost:8800',
                                'type' => 'a.madeup.test.state',
                                'content' => {
                                   'my_key' => 'before'
                                 },
                                'unsigned' => {
                                    'age' => bless( do{\(my $o = 251)}, 'JSON::number' )
                                  }
                              }
                            ]
                          }
                      }
          }
         },
    'next_batch' => 's99_10_0_1_1_1_1_12_1'
  };

```



---

Reference for Dendite

```
docker run --rm --name sytest-dendrite -v "/Users/eric/Documents/github/sytest:/sytest" \
  -v "/Users/eric/Documents/github/dendrite:/src" \
  -v "/Users/eric/logs:/logs" \
  -v "/Users/eric/go:/gopath" -e "POSTGRES=1" -e "DENDRITE_TRACE_HTTP=1" \
  matrixdotorg/sytest-dendrite:latest tests/31sync/09archived.pl
```

Reply:
Here are the events when running the Sytest on Dendrite. I've fixed it so it's including and excluding the proper events now but I think the `join` code that I copied is wrong. It specifically says [`cut all events earlier than the join (but not the join itself)`](https://github.com/matrix-org/dendrite/blob/a85051ca9de82756a94f1ea6870de9bd140a6b20/syncapi/storage/shared/syncserver.go#L888) but then includes the `join` event since an array go slice is inclusive for the start element and exclusive for the end element.

```
0 {event_id:$8QG_YZttke8-VWS-eutTUPd9qvZsZM0RmpSyIhV97U4, sender:@anon-20201231_224222-10:localhost:8800, eventType:m.room.create, origin_server_ts:1609454547251, content:{"creator":"@anon-20201231_224222-10:localhost:8800","room_version":"6"}}, 

1 {event_id:$1q0YVLekwRO2v2BdfMbINGPSunfEIc1Qd5Zr3CSG248, sender:@anon-20201231_224222-10:localhost:8800, eventType:m.room.member, origin_server_ts:1609454547251, content:{"displayname":"anon-20201231_224222-10","membership":"join"}}, 

2 {event_id:$LDkF-AnhrxpQmAb48a_05CIbfvsPuCxRcvHYhQyN1c4, sender:@anon-20201231_224222-10:localhost:8800, eventType:m.room.power_levels, origin_server_ts:1609454547251, content:{"ban":50,"events":{"m.room.aliases":0,"m.room.avatar":50,"m.room.canonical_alias":50,"m.room.history_visibility":100,"m.room.name":50,"m.room.power_levels":100},"events_default":0,"invite":0,"kick":50,"notifications":{"room":50},"redact":50,"state_default":50,"users":{"@anon-20201231_224222-10:localhost:8800":100},"users_default":0}}, 

3 {event_id:$oSEeM7EcdtvlH-s4QrLCwA4Z_0bxxUk3k7_aPp4Pkms, sender:@anon-20201231_224222-10:localhost:8800, eventType:m.room.join_rules, origin_server_ts:1609454547251, content:{"join_rule":"public"}},

4 {event_id:$9L07c7Ii0z0A62McZ_sLBtWD-ygwOJnMjexhFqovTLM, sender:@anon-20201231_224222-10:localhost:8800, eventType:m.room.history_visibility, origin_server_ts:1609454547251, content:{"history_visibility":"shared"}},

5 {event_id:$db4146yqE3o5d-dOrchSk5U3ezzsAhG3zTiL1hk-sBU, sender:@anon-20201231_224222-11:localhost:8800, eventType:m.room.member, origin_server_ts:1609454547426, content:{"avatar_url":"","displayname":"anon-20201231_224222-11","membership":"join"}},

6 {event_id:$xs75AQ_8lw13Qv1mQkWWf3j1PC7dNEvBJIK-OpZvWIw, sender:@anon-20201231_224222-10:localhost:8800, eventType:m.room.message, origin_server_ts:1609454547481, content:{"body":"before","msgtype":"m.text"}},

7 {event_id:$Z8VVPKTPDjQfOLy0avfzAVrqA1HZmltvQo1faNuYxSM, sender:@anon-20201231_224222-10:localhost:8800, eventType:a.madeup.test.state, origin_server_ts:1609454547498, content:{"my_key":"before"}},

8 {event_id:$NWZyf0YuopRAQTpjj2w9Zlw4lJlhTDZcWEK9tOqUV1I, sender:@anon-20201231_224222-11:localhost:8800, eventType:m.room.member, origin_server_ts:1609454547521, content:{"membership":"leave"}},

9 {event_id:$etBSOkdXHwn8jbYlstDP1K-eHH1uZCC_wPb07IEvLBg, sender:@anon-20201231_224222-10:localhost:8800, eventType:m.room.message, origin_server_ts:1609454547542, content:{"body":"after","msgtype":"m.text"}},

10 {event_id:$Q3Wa8HulX7yONb2fIkClP5an02xsiuWqFFS7jxS3l1c, sender:@anon-20201231_224222-10:localhost:8800, eventType:a.madeup.test.state, origin_server_ts:1609454547564, content:{"my_key":"after"}}]"
```

Reply:
Created https://github.com/matrix-org/gomatrixserverlib/issues/245 to track making it easier to debug and dumping out events.